CONTRIBUTING.md

```diff
@@ -98,8 +98,82 @@ object provides an interface to a running Prometheus instance. This
 interface is limited to only those aspects of Prometheus required by
 this charm.
 
+### Library Details
+
+The Prometheus charm library exposes a provider and consumer object -
+`PrometheusProvider` and `PrometheusConsumer` along with the custom
+charm event `TargetsChanged` within the `MonitoringEvents` event
+descriptor. `PrometheusProvider` emits the `TargetsChagned` event in
+response to relation changed and departed events. It is expected that
+the provider charm would respond to this event and regenerate the
+Prometheus configuration using information provided by the `jobs()`
+and `alerts()` methods of the `PrometheusProvider`.
+
+The `jobs()` method gathers a list of scrape jobs from all related
+consumer charms. In doing so it invokes the `_static_scrape_config()`
+method for each such relation. `_static_scrape_config()` generates a
+unique job name for each job, sanitizes the job and associates Juju
+topology labels with it. Labeling of Juju topology is done by
+`_labeled_static_job_config()`. However the labeling differs for
+scrape targets whose host address was automatically gathered and those
+were addresses were explicitly specified. Hence three separate
+functions are required for labeling `_set_juju_labels()` for labels
+common to both but `_labeled_unitless_config()` and
+`_labeled_unit_config()` for the differing labels. The list of
```

Comment:
Consider simplifying

Reply:
You mean simplifying the wording ?

Reply:
I mean increase cognitive ease. It is difficult for me to read and I don't really understand it.

Reply:
This is only a very high level bird eye view of the codebase. It can not be a substitute for the code. It is intendend to facilitate reading the code by giving a lay of the land - what is there and where is it. 

If you have a more specific issue I can address it.